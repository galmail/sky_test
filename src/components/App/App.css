.dashboard {

}

.dashboard > h1 {
  text-align: center;
}