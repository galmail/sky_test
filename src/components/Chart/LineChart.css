/* LineChart.css */

.line-chart {
  float: left;
  width: 760px;
  height: 300px;
  text-align: center;
  line-height: 100px;
}

.line-chart .line {
  fill: none;
  stroke: steelblue;
  stroke-width: 2px;
}

.line-chart .focus circle {
  fill: #F1F3F3;
  stroke: rgb(0, 0, 0);
  stroke-width: 5px;
}
