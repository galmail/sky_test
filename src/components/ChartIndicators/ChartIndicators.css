.chart-indicators {
    float: left;
    border: 1px solid gray;
    border-radius: 5px;
    padding: 10px;
}

.chart-indicators > div {
    margin-bottom: 5px;
}

.chart-indicators > div:hover {
    background-color: yellow;
}