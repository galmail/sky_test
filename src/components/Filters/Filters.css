.filters {
    clear: both;
    overflow: hidden;
    margin-bottom: 20px;
}

.filters > div {
    float: left;
    margin: 0 30px;
}

.filters label {
    margin-right: 10px; 
}

