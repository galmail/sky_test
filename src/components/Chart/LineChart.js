import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as d3 from 'd3/build/d3';

import './LineChart.css';

class LineChart extends Component {

    constructor(props){
        super(props);
        this.rootElement = null;
        this.highlightIndicator = this.highlightIndicator.bind(this);
        this.renderChart = this.renderChart.bind(this);
    }

    render() {
        return (
            <div ref={(el) => { this.rootElement = el; }} className="line-chart">
                The Chart is Loading...
            </div>
        );
    }

    componentDidUpdate() {
        // console.log('LineChart componentDidUpdate', this.props.data);
        if (this.props.data && this.props.data.length > 0) this.renderChart(this.props.data);
        else this.rootElement.innerHTML = 'No data to display';
    }

    renderChart(data) {
        this.rootElement.innerHTML = '';
        if (!data) return;

        // set the dimensions and margins of the chart
        const margin = this.props.margin;
        const width = this.props.width - margin.left - margin.right;
        const height = this.props.height - margin.top - margin.bottom;

        const totalWidth = width + margin.left + margin.right;
        const totalHeight = height + margin.top + margin.bottom;

        // set the ranges
        const x = d3.scaleTime().range([0, width]);
        const y = d3.scaleLinear().range([height, 0]);

        // define the line
        const line = d3.line()
            .x(d => x(d.date))
            .y(d => y(d.value));

        const svg = d3.select(this.rootElement).append("svg")
            .attr("width", totalWidth)
            .attr("height", totalHeight)
            .append("g")
            .attr("transform",
            "translate(" + margin.left + "," + margin.top + ")");

        // Scale the range of the data
        x.domain(d3.extent(data, function (d) { return d.date; }));
        y.domain([0, d3.max(data, function (d) { return d.value; })]);

        // Add the valueline path.
        svg.append("path")
            .data([data])
            .attr("class", "line")
            .attr("d", line);

        // Add the X Axis
        svg.append("g")
            .attr("class", "axis-x")
            .attr("transform", "translate(0," + height + ")")
            .call(d3.axisBottom(x));

        // Add the Y Axis
        svg.append("g")
            .attr("class", "axis-y")
            .call(d3.axisLeft(y));

        // Add Circle
        const focus = svg.append("g")
            .attr("class", "focus")
            .style("display", "none");

        focus.append("circle").attr("r", 7.5);

        this.highlightIndicator(focus, x, y);
    }

    highlightIndicator(el, x, y) {
        const dataObj = this.props.highlightValue;
        const display = dataObj ? null : 'none';
        if (dataObj) el.attr("transform", "translate(" + x(dataObj.date) + "," + y(dataObj.value) + ")");
        el.style("display", display);
    }

}

LineChart.propTypes = {
    margin: PropTypes.shape({
        top: PropTypes.number,
        right: PropTypes.number,
        bottom: PropTypes.number,
        left: PropTypes.number
    }),
    width: PropTypes.number,
    height: PropTypes.number,
    data: PropTypes.arrayOf(PropTypes.shape({
        date: PropTypes.instanceOf(Date),
        value: PropTypes.number
    })),
    highlightValue: PropTypes.shape({
        date: PropTypes.instanceOf(Date),
        value: PropTypes.number
    })
};

LineChart.defaultProps = {
    margin: { top: 20, right: 20, bottom: 30, left: 50 },
    width: 760,
    height: 300
}

export default LineChart;
