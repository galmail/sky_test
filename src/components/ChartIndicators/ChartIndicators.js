import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './ChartIndicators.css';

class ChartIndicators extends Component {
  render() {
    const minTemp = this.props.min != null ? this.props.min : '-----';
    const maxTemp = this.props.max != null ? this.props.max : '-----';
    const avgTemp = this.props.avg != null ? this.props.avg : '-----';
    return (
      <div className="chart-indicators">
        <div className="min-temperature" onMouseOver={() => this.props.onHover('min')} onMouseOut={() => this.props.onHover()}>Min Temp: {minTemp}</div>
        <div className="max-temperature" onMouseOver={() => this.props.onHover('max')} onMouseOut={() => this.props.onHover()}>Max Temp: {maxTemp}</div>
        <div className="avg-temperature" onMouseOver={() => this.props.onHover('avg')} onMouseOut={() => this.props.onHover()}>Avg Temp: {avgTemp}</div>
      </div>
    );
  }
}

ChartIndicators.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  avg: PropTypes.number,
  onHover: PropTypes.func,
};

ChartIndicators.defaultProps = {
  onHover: () => {}
};

export default ChartIndicators;
