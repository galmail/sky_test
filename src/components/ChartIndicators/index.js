import ChartIndicators from './ChartIndicators';
export default ChartIndicators;