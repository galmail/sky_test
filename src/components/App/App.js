import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { getTemperatures } from '../../actions/service';

import LineChart from '../Chart/LineChart';
import Filters from '../Filters';
import ChartIndicators from '../ChartIndicators';

import './App.css';

class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      fromDate: null,
      toDate: null,
      highlightValue: null
    }
    if (props.getTemperatures) props.getTemperatures();
    this.filtersChanged = this.filtersChanged.bind(this);
    this.highlight = this.highlight.bind(this);
  }

  filtersChanged({ fromDate, toDate }) {
    this.setState({ fromDate, toDate });
  }

  highlight({ data, min, max, indicator }){
    let highlightValue = null;
    if (indicator === 'min') {
      highlightValue = data.find(obj => obj.value === min);
    }
    if (indicator === 'max') {
      highlightValue = data.find(obj => obj.value === max);
    }
    this.setState({ highlightValue });
  }

  filterByDates(data, from, to) {
    if (!data || data.length === 0) return data;
    return data.filter(obj => (from ? obj.date >= from : true) && (to ? obj.date <= to : true));
  }

  calculateMinMaxAvg(data) {
    if (!data || data.length === 0) return {};
    const min = data.reduce((prev, curr) => prev.value < curr.value ? prev : curr);
    const max = data.reduce((prev, curr) => prev.value > curr.value ? prev : curr);
    const avg = data.reduce((sum, dataObj) => sum + parseFloat(dataObj.value), 0) / data.length;
    return {
      min: min.value,
      max: max.value,
      avg: Math.round(avg * 100) / 100
    };
  }
  
  render() {
    const data = this.filterByDates(this.props.temperatures, this.state.fromDate, this.state.toDate);
    const { min, max, avg } = this.calculateMinMaxAvg(data);
    return (
      <div className="dashboard">
        <h1>Dashboard</h1>
        <Filters onChange={this.filtersChanged} />
        <LineChart  data={data}
                    highlightValue={this.state.highlightValue}
        />
        <ChartIndicators  min={min}
                          max={max}
                          avg={avg}
                          onHover={(indicator) => this.highlight({ data, min, max, indicator })}
        />
      </div>
    );
  }

}

const mapStateToProps = state => {
  return state;
};

const mapDispatchToProps = dispatch => bindActionCreators({ getTemperatures }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);

export { App };
