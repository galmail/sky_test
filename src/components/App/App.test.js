import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import moment from 'moment';
import ReactTestUtils from 'react-dom/test-utils';

import {Â App } from './App';
import Filters from '../Filters';
import LineChart from '../Chart/LineChart';
import ChartIndicators from '../ChartIndicators';

describe('Test App Component', () => {

  const DATETIME_FORMAT = 'YYYY-MM-DDThh:mm';
  const data = [
    { date: moment('2017-12-10T10:00', DATETIME_FORMAT).toDate(), value: 30.0 },
    { date: moment('2017-12-10T11:00', DATETIME_FORMAT).toDate(), value: 25.0 },
    { date: moment('2017-12-10T12:00', DATETIME_FORMAT).toDate(), value: 10.0 },
  ];

  it('should render without crashing', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Filters).length).toEqual(1);
    expect(wrapper.find(LineChart).length).toEqual(1);
    expect(wrapper.find(ChartIndicators).length).toEqual(1);
    // expect(wrapper.contains(<div className="filters">Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.</div>)).toBe(true);
    // expect(wrapper.contains(<div className="time">February 9, 2017 4:27 AM</div>)).toBe(true);
    // expect(wrapper.contains(<img alt={name} src={user.avatar} />)).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should calculate min,max,avg value given a set of data', () => {
    const wrapper = shallow(<App />);
    const app = wrapper.instance();
    const result = app.calculateMinMaxAvg(data);
    expect(result.min).toEqual(10);
    expect(result.max).toEqual(30);
    expect(result.avg).toEqual(21.67);
  });

  it('should filter the data by dates correctly', () => {
    const wrapper = shallow(<App />);
    const app = wrapper.instance();
    const from = moment('2017-12-10T11:00', DATETIME_FORMAT).toDate();
    const to = new Date();
    const filteredData = app.filterByDates(data, from, to);
    expect(filteredData.length).toBe(2);
    expect(filteredData[0].value).toBe(25);
  });

  it('should update chart when filters change', () => {
    const wrapper = mount(<App />);
    // const wrapper = shallow(<App />);
    const app = wrapper.instance();
    app.filtersChanged = jest.fn();
    const startDate = wrapper.find('#startDateId').get(0);
    // startDate.props.value = '2017-12-10T12:00';
    // const filters = wrapper.find(Filters);
    // ReactTestUtils.Simulate.change(filters);
    // expect(app.filtersChanged.mock.calls.length).toBe(1);

  });

  it('should highlight minValue on chart when highlight is called', () => {

  });


});

