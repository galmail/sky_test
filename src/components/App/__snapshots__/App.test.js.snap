exports[`Test App Component should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Dashboard
</h1>,
        <Filters
          fromDate={2016-12-11T00:03:11.537Z}
          onChange={[Function]}
          toDate={2017-12-11T00:03:50.150Z} />,
        <LineChart
          data={undefined}
          height={300}
          highlightValue={null}
          margin={
                    Object {
                              "bottom": 30,
                              "left": 50,
                              "right": 20,
                              "top": 20,
                            }
          }
          width={760} />,
        <ChartIndicators
          avg={undefined}
          max={undefined}
          min={undefined}
          onHover={[Function]} />,
      ],
      "className": "dashboard",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Dashboard",
        },
        "ref": null,
        "rendered": "Dashboard",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "fromDate": 2016-12-11T00:03:11.537Z,
          "onChange": [Function],
          "toDate": 2017-12-11T00:03:50.150Z,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data": undefined,
          "height": 300,
          "highlightValue": null,
          "margin": Object {
            "bottom": 30,
            "left": 50,
            "right": 20,
            "top": 20,
          },
          "width": 760,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "avg": undefined,
          "max": undefined,
          "min": undefined,
          "onHover": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Dashboard
</h1>,
          <Filters
            fromDate={2016-12-11T00:03:11.537Z}
            onChange={[Function]}
            toDate={2017-12-11T00:03:50.150Z} />,
          <LineChart
            data={undefined}
            height={300}
            highlightValue={null}
            margin={
                        Object {
                                    "bottom": 30,
                                    "left": 50,
                                    "right": 20,
                                    "top": 20,
                                  }
            }
            width={760} />,
          <ChartIndicators
            avg={undefined}
            max={undefined}
            min={undefined}
            onHover={[Function]} />,
        ],
        "className": "dashboard",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Dashboard",
          },
          "ref": null,
          "rendered": "Dashboard",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "fromDate": 2016-12-11T00:03:11.537Z,
            "onChange": [Function],
            "toDate": 2017-12-11T00:03:50.150Z,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data": undefined,
            "height": 300,
            "highlightValue": null,
            "margin": Object {
              "bottom": 30,
              "left": 50,
              "right": 20,
              "top": 20,
            },
            "width": 760,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "avg": undefined,
            "max": undefined,
            "min": undefined,
            "onHover": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
